File: labs-opensource/M06-multi-tool-orchestration/starter/research_agent.py
"""
M06 Lab - Step 2: The Orchestrating Loop
=========================================
The M05 loop + tool filtering + max-iterations guard + PARALLEL execution.
Run: python research_agent.py
"""

import json
from concurrent.futures import ThreadPoolExecutor, as_completed

from openai import OpenAI
from tools_registry import build_registry, execute_tool

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")
registry = build_registry()

SYSTEM = (
    "You are a research assistant. When asked to compare or research multiple "
    "topics, search for each one in parallel. When asked to fetch and summarize "
    "a page, do it sequentially."
)


def run_agent(question: str, tool_tags: list[str] = None, verbose: bool = True) -> str:
    """Run the multi-tool agent. Optionally filter tools by tag.

    TODO:
    1. active_tools = registry.get_tools_for_context(tags=tool_tags) if tool_tags
       else registry.get_tools_for_context()
    2. messages = [{"role": "system", "content": SYSTEM},
                   {"role": "user", "content": question}]
    3. for iteration in range(10):          ← max-iterations guard, NOT while True
       a. response = client.chat.completions.create(model="mistral",
              tools=active_tools, messages=messages)
          (try/except → return f"API error: {e}")
       b. choice = response.choices[0]; tool_calls = choice.message.tool_calls or []
       c. If choice.finish_reason == "stop" or not tool_calls:
            return choice.message.content or ""
       d. Append the assistant message (content + tool_calls) to messages
       e. If verbose: print "PARALLEL" if len(tool_calls) > 1 else "SEQUENTIAL"
          and each requested call
       f. EXECUTE:
          - If len(tool_calls) > 1: use ThreadPoolExecutor + as_completed —
            submit execute_tool(name, json.loads(arguments)) per call, map each
            future back to its tool_call.id, build the {"role": "tool", ...} list
          - Else: run the single call directly
          REMEMBER: every tool_call_id needs a result message, including errors
       g. messages.extend(tool_results) and loop back
    4. After the loop: return "Max iterations reached."
    """
    pass  # Remove this line when you add your code


# ── Test Scenarios (COMPLETE) ──
if __name__ == "__main__":
    print("\n" + "> TEST 1: PARALLEL SEARCH ".ljust(60, "-"))
    r1 = run_agent(
        "Search for information about these 3 topics: AI agents, "
        "prompt engineering, and tool use patterns.",
        tool_tags=["research"],
    )
    print(f"\nResult preview: {(r1 or '')[:200]}...")

    print("\n" + "> TEST 2: SEQUENTIAL CHAIN ".ljust(60, "-"))
    r2 = run_agent(
        "Search for 'Mistral AI tool use', then fetch the first "
        "result page and summarize its content.",
        tool_tags=["research"],
    )
    print(f"\nResult preview: {(r2 or '')[:200]}...")

    print("\n" + "> TEST 3: ERROR RECOVERY ".ljust(60, "-"))
    r3 = run_agent(
        "Fetch and summarize this page: https://broken.example.com/404",
        tool_tags=["research"],
    )
    print(f"\nResult preview: {(r3 or '')[:200]}...")

    print("\n" + "> TEST 4: DYNAMIC TOOL FILTERING ".ljust(60, "-"))
    r4 = run_agent(
        "Format a citation for an article titled 'Multi-Tool AI Agents' "
        "from https://example.com/agents, accessed today.",
        tool_tags=["citation"],
    )
    print(f"\nResult preview: {(r4 or '')[:200]}...")

File: labs-opensource/M06-multi-tool-orchestration/starter/research_agent.js
/**
 * M06 Lab - Step 2: The Orchestrating Loop
 * =========================================
 * The M05 loop + tool filtering + max-iterations guard + PARALLEL execution.
 * Run: node research_agent.js
 */

import OpenAI from "openai";
import { buildRegistry, executeTool } from "./tools_registry.js";

const client = new OpenAI({ baseURL: "http://localhost:11434/v1", apiKey: "ollama" });
const registry = buildRegistry();

const SYSTEM =
  "You are a research assistant. When asked to compare or research multiple " +
  "topics, search for each one in parallel. When asked to fetch and summarize " +
  "a page, do it sequentially.";

/**
 * Run the multi-tool agent. Optionally filter tools by tag.
 *
 * TODO:
 * 1. const activeTools = toolTags
 *      ? registry.getToolsForContext({ tags: toolTags })
 *      : registry.getToolsForContext();
 * 2. const messages = [{ role: "system", content: SYSTEM },
 *                      { role: "user", content: question }];
 * 3. for (let iteration = 0; iteration < 10; iteration++) {   ← guard, NOT while(true)
 *    a. response = await client.chat.completions.create({ model: "mistral",
 *         tools: activeTools, messages })   (try/catch → return `API error: ...`)
 *    b. const choice = response.choices[0];
 *       const toolCalls = choice.message.tool_calls ?? [];
 *    c. If choice.finish_reason === "stop" || toolCalls.length === 0:
 *         return choice.message.content ?? "";
 *    d. Push the assistant message (content + tool_calls) onto messages
 *    e. If verbose: log "PARALLEL" if toolCalls.length > 1 else "SEQUENTIAL"
 *    f. EXECUTE:
 *       - If toolCalls.length > 1: Promise.all over the calls — each resolves to
 *         { role: "tool", tool_call_id: tc.id, content: result }
 *       - Else: await the single call
 *       REMEMBER: every tool_call_id needs a result message, including errors
 *    g. messages.push(...toolResults) and loop back
 * }
 * 4. After the loop: return "Max iterations reached.";
 */
async function runAgent(question, toolTags = null, verbose = true) {
  // TODO: implement
}

// ── Test Scenarios (COMPLETE) ──
console.log("\n> TEST 1: PARALLEL SEARCH " + "-".repeat(34));
const r1 = await runAgent(
  "Search for information about these 3 topics: AI agents, prompt engineering, and tool use patterns.",
  ["research"]
);
console.log(`\nResult preview: ${(r1 ?? "").slice(0, 200)}...`);

console.log("\n> TEST 2: SEQUENTIAL CHAIN " + "-".repeat(33));
const r2 = await runAgent(
  "Search for 'Mistral AI tool use', then fetch the first result page and summarize its content.",
  ["research"]
);
console.log(`\nResult preview: ${(r2 ?? "").slice(0, 200)}...`);

console.log("\n> TEST 3: ERROR RECOVERY " + "-".repeat(35));
const r3 = await runAgent(
  "Fetch and summarize this page: https://broken.example.com/404",
  ["research"]
);
console.log(`\nResult preview: ${(r3 ?? "").slice(0, 200)}...`);

console.log("\n> TEST 4: DYNAMIC TOOL FILTERING " + "-".repeat(27));
const r4 = await runAgent(
  "Format a citation for an article titled 'Multi-Tool AI Agents' from https://example.com/agents, accessed today.",
  ["citation"]
);
console.log(`\nResult preview: ${(r4 ?? "").slice(0, 200)}...`);
